### PR TITLE
added function to model HCD following Rogers et al. 2018

### PR DESCRIPTION
**Pull request by andreufont**
I was able to run fitter2 with this model, and I get L = 10 Mpc/h, but I get a worse chi2 than with the sinc function. I haven't been able to make plots for both models, but it would be interesting to compare where they disagree.

Is there anything extra I should do before merging?

### CODE REVIEW

**Comment by ngbusca** [comment]
@andreufont I thought `b` was a free parameter in Rogers' model? Wouldn't we want to leave it free even if it isn't?

File: py/picca/fitter2/pk.py [hunk 1]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
 
     return pk
 
+def pk_hcd_Rogers2018(k, pk_lin, tracer1, tracer2, **kwargs):
+    # alternative to pk_hcd with a physically motivated model (Rogers++ 2018)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: Please add docstrings like in https://github.com/igmhub/picca/blob/5c44d4ecfcae7c25e79b7e5ed84f6172e2cd287a/py/picca/utils.py#L141.
Add the equation number, the page and the arxiv id of the paper.

File: py/picca/fitter2/pk.py [hunk 2]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+
+    bias1, beta1, bias2, beta2 = bias_beta(kwargs, tracer1, tracer2)
+
+    bias_hcd = kwargs["bias_hcd"]
+    beta_hcd = kwargs["beta_hcd"]
+    L0 = kwargs["L0_hcd"]
+
+    kp = k*muk
+    # Lorentzian part
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: no need for comment

File: py/picca/fitter2/pk.py [hunk 3]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+    F_hcd = utils.sinc(kp*L0)
+    # Gaussian part (1 Mpc/h ~ 100 km/s)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: no need for comment

File: py/picca/fitter2/pk.py [hunk 4]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+    b = 1.0
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: why isn't that free?

File: py/picca/fitter2/pk.py [hunk 5]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+    F_hcd = sp.exp(-L0*kp) * sp.exp(-0.5*kp*kp*b*b)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: `(kp*b)**2`

File: py/picca/fitter2/pk.py [hunk 6]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+
+    bias_eff1 = (bias1 + bias_hcd*F_hcd)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: useless parenthesis

File: py/picca/fitter2/pk.py [hunk 7]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+    beta_eff1 = (bias1 * beta1 + bias_hcd*beta_hcd*F_hcd)/(bias1 + bias_hcd*F_hcd)
+
+    bias_eff2 = (bias2 + bias_hcd*F_hcd)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: useless parenthesis

File: py/picca/fitter2/pk.py [hunk 8]
```diff
@@ -52,6 +52,32 @@ def pk_hcd(k, pk_lin, tracer1, tracer2, **kwargs):
...
+    beta_eff2 = (bias2 * beta2 + bias_hcd*beta_hcd*F_hcd)/(bias2 + bias_hcd*F_hcd)
+
+    pk = pk_lin*bias_eff1*bias_eff2*(1 + beta_eff1*muk**2)*(1 + beta_eff2*muk**2)
```
Comment (londumas) [inline py/picca/fitter2/pk.py]: `1.+`

**Comment by londumas** [comment]
@andreufont, it would be nice to have a plots of the two best fit models and data, on the four wedges.

File: py/picca/fitter2/pk.py (hunk 1 above)
Comment (andreufont) [reply py/picca/fitter2/pk.py]: None of the functions in this file have these docstrings...

File: py/picca/fitter2/pk.py (hunk 3 above)
Comment (andreufont) [reply py/picca/fitter2/pk.py]: no comment

File: py/picca/fitter2/pk.py (hunk 4 above)
Comment (andreufont) [reply py/picca/fitter2/pk.py]: (see reply to Nico below)

File: py/picca/fitter2/pk.py (hunk 1 above)
Comment (andreufont) [reply py/picca/fitter2/pk.py]: I added a docstring with what I know (I don't know the input/output format of pk_hcd).

**Comment by ngbusca** [comment]
@andreufont I would leave `b` as a free parameter and fix it in the `. ini`

for the docstring, it's your function that makes assumptions on what `pk` or other args are and the docstring should reflext that. E. g. I guess in your case it should be a numpy array...

**Comment by londumas** [comment]
@ngbusca, the new PR is perfect. Could you plot the effect on the fit?

**Comment by andreufont** [comment]
@londumas - I'm trying to figure out how to use the h5 files, I think I'll figure it out before 2019 :-)

**Comment by ngbusca** [comment]
@andreufont you can take a look at the `tutorials/QA_plots.ipynb`. If you will make the plots on a notebook, don't hesitate PRing them in!

**Comment by andreufont** [comment]
> @andreufont you can take a look at the `tutorials/QA_plots.ipynb`. If you will make the plots on a notebook, don't hesitate PRing them in!

@ngbusca - yes, Helion pointed me to that, but it was very difficult to identify the relevant lines. He sent me a different shorter script, and after a while I manage to get the plots :-)

I'm running now a proper fitting with metals, and will post the results.

**Comment by andreufont** [comment]
The degeneracies between L_HCD and beta_F make it very difficult to do these comparisons. I ended up generating plots with L_HCD fixed at different values based on the fits done by Keir Rogers on the simulated DLAs. For instance, L6 in the mocks means a Lorentzian model with L=6 Mpc/h, a reasonable value if the contaminants were dominated by small (or sub-) DLAs. 

wedge_hcd_0 0_0 5
wedge_hcd_0 5_0 8
wedge_hcd_0 8_0 95
wedge_hcd_0 95_1 0
wedge_hcd_zoom

**Comment by andreufont** [comment]
In any case I think we are ready to merge.

**Comment by londumas** [comment]
@andreufont, Thanks for the work.

**Pull request merged**